src/mbr.c

```diff
@@ -0,0 +1,438 @@
+
+	lba /= geometry->sectors;
+	chs->head = lba % geometry->heads;
+
+	lba /= geometry->heads;
+	chs->cylinder = lba & 0xFF;
+
+	/* Move bit 8 & 9 of cylinder to bit 6 & 7 of sector */
+	chs->sector |= (lba >> 2) & 0xC0;
+}
+
+static gboolean get_raw_partition_entry(gint fd,
+		struct partition_tbl_entry *raw_entry,
+		const struct mbr_switch_partition *partition, GError **error)
+{
+	gboolean res = FALSE;
+	guint sector_size;
+	struct hd_geometry geometry;
+	GError *ierror = NULL;
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+	g_assert_nonnull(raw_entry);
+
+	sector_size = get_sectorsize(fd);
+
+	if (partition->start % sector_size || partition->size % sector_size) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Partition start address or size is not a multipe"
+				" of sector size %d", sector_size);
+		goto out;
+	}
+
+	res = get_hd_geometry(fd, &geometry, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	raw_entry->partition_start = partition->start / sector_size;
+	raw_entry->partition_size = partition->size / sector_size;
+
+	get_chs(&raw_entry->chs_start, raw_entry->partition_start, &geometry);
+
+	get_chs(&raw_entry->chs_end,
+			raw_entry->partition_start + raw_entry->partition_size - 1,
+			&geometry);
+out:
+	return res;
+}
+
+gboolean r_mbr_switch_get_inactive_partition(const gchar *device,
+		struct mbr_switch_partition *partition,
+		guint64 region_start, guint64 region_size,
+		GError **error)
+{
+	gboolean res = FALSE;
+	struct mbr mbr;
+	GError *ierror = NULL;
+	struct partition_tbl_entry *boot_part;
+	guint sector_size;
+	gint fd;
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	fd = g_open(device, O_RDONLY);
+
+	sector_size = get_sectorsize(fd);
+
+	res = validate_region(fd, region_start, region_size, sector_size, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	res = read_mbr(fd, &mbr, &ierror);
+	if (!res) {
+		g_propagate_prefixed_error(error, ierror,
+				"failed to read mbr:");
+		goto out;
+	}
+
+	/* check if region overlaps with any partition */
+	res = is_region_free(region_start, region_size, mbr.partition_table,
+			sector_size, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+	res = FALSE;
+
+	boot_part = &mbr.partition_table[BOOT_PARTITION_ENTRY];
+
+	if (boot_part->partition_start == 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"no boot partition found in entry %d",
+				BOOT_PARTITION_ENTRY);
+		goto out;
+	}
+
+	if ((region_start / sector_size) ==
+	    (guint64)boot_part->partition_start) {
+		partition->start = region_start + region_size / 2;
+	}
+	else if (((region_start + region_size / 2) / sector_size) ==
+	         (guint64)boot_part->partition_start) {
+		partition->start = region_start;
+	}
+	else {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"boot partition's start address does match region "
```

Comment:
I assume this should be 'doesn't' ? Message should also start with a capital letter to be consistent with others.